### ISSUE TITLE
Integrating Fex support

### ISSUE THREAD

**Issue by rowanG077** [tpwrules/nixos-apple-silicon#237]
With fex + vulkan seeming to be pretty close on fedora asahi I think we should start thinking about how we can  support fex on nixos. I personally have little experience how rootfs functions so this is more to get the discussion started. I'm willing to work on this so if anyone has any pointers to achieve this that would help as well.

**Comment by rowanG077** [comment]
As a starting point, the following expression builds fex without tests. Tests fail due to a linking problem. I can run FEXconfig without problems. I choose master branch instead of last release because FEXConfig didn't work yet with 16k pagesize on the release of last month. 

```
{ pkgs ? import <nixpkgs> {} }:

let
  inherit (pkgs)
    clangStdenv fetchgit cmake clang ninja pkg-config lld llvm openssl
    autoPatchelfHook nasm squashfsTools squashfuse erofs-utils qt5 python3 python3Packages;

in clangStdenv.mkDerivation rec {
  pname = "fex-emu";
  version = "master";

  src = fetchgit {
    url = "https://github.com/FEX-Emu/FEX.git";
    rev = "5e9c2110dbd26b2bd92a3fbf7221c12d368ee655";
    sha256 = "sha256-3bYC49EcE0X1JjDUlsAIVT4uIVdG+SzbyM7YJDDh+Jk=";
    fetchSubmodules = true;
  };

  nativeBuildInputs = [
    cmake
    clang
    ninja
    pkg-config
    lld
    nasm
    llvm
    python3
    python3Packages.setuptools
    python3Packages.libclang
    autoPatchelfHook
    qt5.wrapQtAppsHook
  ];

  buildInputs = [
    pkgs.stdenv.cc.cc.lib
    openssl.dev
    squashfsTools
    squashfuse
    erofs-utils
    qt5.full
  ];

  cmakeFlags = [
    "-DUSE_LINKER=lld"
    "-DENABLE_LTO=True"
    "-DENABLE_ASSERTIONS=FALSE"
    "-DOVERRIDE_VERSION=FEX-${version}"
    "-DBUILD_TESTS=FALSE"
  ];

  meta = with pkgs.lib; {
    description = "FEX - Fast x86 emulation frontend";
    homepage = "https://github.com/FEX-Emu/FEX";
    license = licenses.mit;
    platforms = platforms.linux;
  };
}
```

The expression is rough and probably some dependency can be removed.

**Comment by rowanG077** [comment]
Pinging @RossComputerGuy as you have packaged krun aka muvm upstream. Currently it crashes for me but maybe that is known.

**Comment by RossComputerGuy** [comment]
This is definitely a good idea and I can look into it when I have a moment. (I should rename krun to muvm in nixpkgs, didn't know about the rename until recently)

**Comment by rowanG077** [comment]
Perhaps it's my config but when I try to run `krun`/`muvm` on my machine which tracks `nixos-unstable` I get a panic:

```
[nix-shell:~]$ krun echo hello
thread '<unnamed>' panicked at src/vmm/src/builder.rs:412:49:
called `Result::unwrap()` on an `Err` value: TooLarge
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: Rust cannot catch foreign exceptions
Aborted (core dumped)
```

**Comment by rowanG077** [comment]
This is a fex derivation, I can't test it because you need a 4K pagesize system to run it. For which we will need muvm. If you disable tests though it fully builds and you can at the very least run `FEXConfig`. 

```
{ lib, stdenv, clangStdenv, fetchgit, cmake, clang, ninja, pkg-config, lld, llvm
, llvmPackages, openssl, glibc, autoPatchelfHook, nasm, qt5, python, setuptools
, libclang
}:

clangStdenv.mkDerivation rec {
  pname = "fex-emu";
  version = "2410";

  src = fetchgit {
    url = "https://github.com/FEX-Emu/FEX.git";
    rev = "FEX-${version}";
    sha256 = "sha256-btzb7OGa3M89wDn8AK/iocT3GLY1mB3cZ0iuWAyNIFc=";
    fetchSubmodules = true;
  };

  nativeBuildInputs = [
    cmake
    clang
    ninja
    pkg-config
    nasm
    qt5.wrapQtAppsHook
    # required with "-DUSE_LINKER=lld"
    # lld
    # autoPatchelfHook
  ];

  buildInputs = [
    llvm
    python
    setuptools
    libclang
    openssl.dev
    qt5.full
  ];

  cmakeFlags = [
    # upstream recommends this but binaries have linker issues
    # libstdc++.so.6: cannot open shared object file: No such file or directory
    # autoPatchelfHook fixes that for the output binaries but test are run
    # by cmake prior to that
    # "-DUSE_LINKER=lld"
    "-DENABLE_ASSERTIONS=FALSE"
    "-DBUILD_TESTS=FALSE"
    "-DOVERRIDE_VERSION=FEX-${version}"
  ];

  meta = with lib; {
    description = "FEX - Fast x86 emulation frontend";
    homepage = "https://github.com/FEX-Emu/FEX";
    license = licenses.mit;
    platforms = platforms.linux;
    maintainers = with maintainers; [ rowanG077 ];
  };
}
```

**Comment by Lemm1** [comment]
> This is a fex derivation, I can't test it because you need a 4K pagesize system to run it. For which we will need muvm. If you disable tests though it fully builds and you can at the very least run `FEXConfig`.
> 
The problem with FEX is to create RootFS, there are prebuilds for fedora, arch and ubuntu. So for nixos you will need to create your own. I've tried to do it few months ago, but it wasn't that easy as creating FEX package. Also in order to make something distributable it will require some nix wizard to somehow either automate RootFS creation on package build or to have prebuilt RootFS distributed like FEX do it for some distros. I personally gave up on that and started using distrobox for FEX

**Comment by rencire** [comment]
I was doing some basic research and found this post that could be helpful for those who want to pursue this: https://old.reddit.com/r/AsahiLinux/comments/1hd6wkg/fexemu_on_asahi_debian/m1u7uxr/

**Comment by mkurz** [comment]
Hi,

we made fex + muvm + steam work on arch linux arm, now hosted under https://asahi-alarm.org/ (GitHub: https://github.com/asahi-alarm/asahi-alarm/ and the other repos in that org).

To build the rootfs we forked https://github.com/FEX-Emu/RootFS - see how we set up the config in this branch: https://github.com/asahi-alarm/RootFS/tree/asahi-alarm (or look at the [diff](https://github.com/FEX-Emu/RootFS/compare/main...asahi-alarm:RootFS:asahi-alarm))
To build the erofs image we now can just run: `./build_image.py -m 2G -no-repack-tar -no-repack-squashfs Configs/Arch.json cache_dir rootfs_dir` (you might also want to disable kvm if needed: ` -disable-kvm`) on an Ubuntu 24.04 x86_64 machine where we have `apt-file python3-git python3-requests qemu-utils pigz guestfs-tools cloud-image-utils erofs-utils linux-image-kvm` installed.

And here is how we build the overlays: https://github.com/asahi-alarm/mesa-overlay/blob/main/Dockerfile (running with `docker build --no-cache=true --output=build .`) - using the mesa [PKGBUILD](https://github.com/asahi-alarm/PKGBUILDs/blob/main/mesa-asahi/PKGBUILD) - which is 1:1 what Fedora is using (by the means of whichs make flags they are using and which files the install)

Here are the PKGBUILDs for all our packages: https://github.com/asahi-alarm/PKGBUILDs

Maybe this helps, have a nice day!

**Comment by nrabulinski** [comment]
I'm currently running NixOS with Asahi kernel on my M1 MacBook Pro and got muvm with x86 emulation and gpu acceleration working. It's not super clean and sadly x86 emulation doesn't work in the Nix sandbox yet which I consider a must, but I'll be submitting the required patches to muvm first before cleanly packaging the rest

**Comment by rowanG077** [comment]
@nrabulinski Awesome, yeah I saw your issues and discussion on muvm! I look forward to testing your work.

**Comment by nrabulinski** [comment]
In the meantime it'd be awesome if someone could get a cross-compile working to build x86_64-linux version of mesa-asahi-edge on aarch64-linux. I worked around it for my usecase by simply using a native x86_64-linux builder, but that's not sustainable and not everyone will have that luxury. An alternative would be to require people to first "bootstrap" by building muvm and fex first, registering them as x86 interpreter, and then building the graphics driver as an x86_64 system, but that's not very elegant.
Is there a nixos-on-asahi community somewhere already? If not, maybe people would be interested in creating a dedicated matrix room at nixos.org?

**Comment by nrabulinski** [comment]
Update: my PRs to muvm are open and waiting for review. Also, with the help of @yuyuyureka I managed to cross-compile mesa for x86 on my asahi machine. I haven't yet tried packaging fex but we're closer than ever to proper x86 app support on NixOS asahi :) 
The steam wrapper will still have to wait, as well as running x86 programs inside the Nix sandbox, but for day-to-day use it should make nixos with asahi kernel much more usable

**Comment by nrabulinski** [comment]
I've created https://github.com/nrabulinski/nixos-muvm-fex for my muvm+fex experiments, before starting to upstream those efforts.
If someone just wants to run x86 linux apps on their asahi machine, well that's sufficient I'm doing so right now. GUI apps work just fine, with hardware acceleration :)
The current limitations are: you can't use muvm+fex as binfmt handler, even if you could, muvm doesn't like the nix sandbox. You also cannot use nix from inside the muvm guest, that's also work that's in progress as we speak 

We also have official matrix channel, feel free to join at [#asahi:nixos.org](https://matrix.to/#/#asahi:nixos.org)

**Comment by rowanG077** [comment]
You are a true legend @nrabulinski!

**Comment by Yeshey** [comment]
> I've created https://github.com/nrabulinski/nixos-muvm-fex for my muvm+fex experiments, before starting to upstream those efforts. If someone just wants to run x86 linux apps on their asahi machine, well that's sufficient I'm doing so right now. GUI apps work just fine, with hardware acceleration :) The current limitations are: you can't use muvm+fex as binfmt handler, even if you could, muvm doesn't like the nix sandbox. You also cannot use nix from inside the muvm guest, that's also work that's in progress as we speak
> 
> We also have official matrix channel, feel free to join at [#asahi:nixos.org](https://matrix.to/#/#asahi:nixos.org)

Can this be used to run x86 software like steamcmd in a KVM server like Oracle's Ampare A1 arm servers?

**Comment by RossComputerGuy** [comment]
I think so, though it shouldn't be exactly required if you're running with a 4k page size. I've been trying with regular steam on Ampere Altra Max M128-26.

**Comment by Yeshey** [comment]
> I think so, though it shouldn't be exactly required if you're running with a 4k page size. I've been trying with regular steam on Ampere Altra Max M128-26.

cool, yeah, Oracle Ampare A1 does have 4K page size, that means muvm is not needed then, right?
And how are you running steam, or x86 packages? Are you getting them from nixpkgs? I have a project for box64 where I added an overlay to allow installing x86 apps on aarch64 with:
```nix
  imports = [ inputs.box64-binfmt.nixosModules.default ];
  nixpkgs.overlays = [ inputs.box64-binfmt.overlays.default ];

  environment.systemPackages = [
    pkgs.x86.steamcmd
    pkgs.x86.wineWowPackages.stable
    pkgs.x86.katawa-shoujo
  ];
``` 
But that is kinda hacky, and sets up a bunch of things for `qemu` emulation to allow building:
```nix
boot.binfmt.emulatedSystems = ["i686-linux" "x86_64-linux" "i386-linux" "i486-linux" "i586-linux" "i686-linux"];
nix.settings.extra-platforms = ["i686-linux" "x86_64-linux" "i386-linux" "i486-linux" "i586-linux" "i686-linux"];
```
And then when removing it from my config it might happen that it ends up with wrong architecture binaries in the middle of the process and it trips out, couldn't build configuration, and I had to redeploy the server, I don't have time to try my own flake again rn ahahah

idk if I was overengineering ahaha

**Comment by RossComputerGuy** [comment]
I have not gotten Steam working due to muvm failing to pass through the NVIDIA GPU I have. I've also tried getting Steam in but libX11.so.6 fails to load for some reason.

**Comment by rowanG077** [comment]
Now that muvm is upstream I thought I'd try it. I get a panic but it still launches the command.

```sh
> nix-shell -p muvm
[nix-shell:~]$ RUST_BACKTRACE=1 muvm sh

thread 'gpu worker' panicked at src/devices/src/virtio/gpu/virtio_gpu.rs:254:14:
Rutabaga initialization failed!: ComponentError(-19)
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: devices::virtio::gpu::worker::Worker::work
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Using default interface naming scheme 'v257'.
sh-5.2$ /nix/store/vlzzppxvd7swdy67kj3avr5r9147llzf-udev-rules/51-android.rules:1101 Unknown group 'adbusers', ignoring.

sh-5.2$ 
```

Is there anything I need to do to get this working? I'm also interested in running steam.

**Comment by RossComputerGuy** [comment]
That's the same issue I see on my Ampere desktop with an NVIDIA RTX 5070. I don't see that issue on my M1 Pro MBP.

**Comment by rowanG077** [comment]
I'm using an M2 Max for reference. Using this repo + nixos unstable. I use "replace" mesa option if it matters.

**Comment by Liamolucko** [comment]
> Now that muvm is upstream I thought I'd try it. I get a panic but it still launches the command.
> 
> > nix-shell -p muvm
> [nix-shell:~]$ RUST_BACKTRACE=1 muvm sh
> 
> thread 'gpu worker' panicked at src/devices/src/virtio/gpu/virtio_gpu.rs:254:14:
> Rutabaga initialization failed!: ComponentError(-19)
> stack backtrace:
>    0: rust_begin_unwind
>    1: core::panicking::panic_fmt
>    2: core::result::unwrap_failed
>    3: devices::virtio::gpu::worker::Worker::work
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> Using default interface naming scheme 'v257'.
> sh-5.2$ /nix/store/vlzzppxvd7swdy67kj3avr5r9147llzf-udev-rules/51-android.rules:1101 Unknown group 'adbusers', ignoring.
> 
> sh-5.2$ 
> 
> Is there anything I need to do to get this working? I'm also interested in running steam.

That looks like the error I get when I'm not using Asahi's patched version of virglrenderer:

```nix
virglrenderer = prev.virglrenderer.overrideAttrs (old: {
  src = final.fetchurl {
    url = "https://gitlab.freedesktop.org/asahi/virglrenderer/-/archive/asahi-20250424/virglrenderer-asahi-20250424.tar.bz2";
    hash = "sha256-9qFOsSv8o6h9nJXtMKksEaFlDP1of/LXsg3LCRL79JM=";
  };
  mesonFlags = old.mesonFlags ++ [ (final.lib.mesonOption "drm-renderers" "asahi-experimental") ];
})
```

> That's the same issue I see on my Ampere desktop with an NVIDIA RTX 5070. I don't see that issue on my M1 Pro MBP.

I'm not sure why it's working on your M1 Pro (maybe you are using the patched version and forgot about it?), but I think it's expected for it not to be working with an NVIDIA GPU: it seems like muvm only supports GPU acceleration using DRM native context, which isn't supported on NVIDIA GPUs (https://github.com/AsahiLinux/muvm/blob/main/README.md?plain=1#L67).

**Comment by rowanG077** [comment]
@Liamolucko Can confirm that with that patch `virglrenderer` I no longer see the panic. But honestly at this point I'm kind stumped. How can I get an x86 nixos environment here, or perhaps even a fedora environment? Or is that not yet possible?

**Comment by psanford** [comment]
I was able to get steam running along with a few games I have handy (Void War, Enter the Gungeon).

I had to hack things up pretty bad so I don't have a nice set of instructions yet. I will try to cleanup my setup this week and then share out instructions.

In the meantime here's briefly what I did:
- running muvm with the [overlays provided by fedora](https://src.fedoraproject.org/rpms/fex-emu-rootfs-fedora) is sufficient to get Fex working with basic (non-gui) x86 binaries
- The Steam wrapper scripts work in this environment but then steam segfaults and it wasn't clear why. I decided it would be easier to setup a fedora rootfs where I understood better how things should work
- I launched a fedora docker image, installed the asahi yum.repo.d repos, and then installed the asahi specific packages (mesa)
- I ran podman export ... to make a rootfs directory 
- I patched muvm to mount this instead of /. 
- I also patched muvm so that it would work even without / mounted (it makes a lot of assumptions that / is the host's /)
- I passed through  MESA_LOADER_DRIVER_OVERRIDE=asahi
- This was enough to get glxgears working
- I extracted steam from the .deb and ran its install script
- There was as issue with the wrapper script getting hung running cat. I assume this is something very silly but I just killed cat to get it to continue
- Steam worked

**Comment by Zynh0722** [comment]
@psanford I'd absolutely love to put in some leg work in getting something that functions for nix, even if it's not something that fits directly in this repository. Unfortunately I'm not quite familiar enough with Fedora environments to reproduce your steps, and I couldn't find a way to contact you directly haha.

If you'd be willing to help me get through some of the gaps in my knowledge that would be incredible. 

I also couldn't find anything else within this thread that got me anywhere useful, but if there is already something that exists that is largely plug and play for nixos users that people know about that would be great too haha

edit: at this point I just want something that works even a little bit to play games on the go haha

**Comment by psanford** [comment]
Sorry it took a bit for me to get back to this. I cleaned up my muvm patches so they should be usable for others now. I have dropped basic instructions for running this here: https://github.com/psanford/muvm/blob/alt-rootfs/NIXOS-APPLE-SILICON.md

**Comment by neobrain** [comment]
Thanks for writing up instructions! Just curious since I stumbled upon this the other day on Fedora, does your RootFS bundle all locales? It turns out not bundling any of the locale indicated by *any* `LC_` env variable will add a 5 second delay to pressure-vessel, which adds up to a very noticeable slow down in start up Steam itself and games launched from Steam.

If locales aren't provided, `pv-adverb[449]: W: Container startup will be faster if missing locales are created at OS level` will be printed in webhelper-linux.txt and console-linux.txt.

**Comment by slendidev** [comment]
@psanford Hello, thanks for the guide. I got it to kinda work, had to add this flag: `--mount run:/run/user:/run/muvm-host/run/user`. glxgears opens, renders a frame, but then nothing else. Know anything about this? Thank you. Steam also opens up a black window, which I think is supposed to be the login window.

**Comment by psanford** [comment]
Do you have the same mesa version inside of the VM as you do outside?

**Comment by slendidev** [comment]
> Do you have the same mesa version inside of the VM as you do outside?

Nope. On my host I have:
```
OpenGL version string: 4.6 (Compatibility Profile) Mesa 25.2.3
```
And in the fedora rootfs:
```
OpenGL version string: 4.6 (Compatibility Profile) Mesa 25.2.1
```
Know a way to get them to match?

**Comment by rowanG077** [comment]
I followed the tutorial and `glxinfo` and `glxgears` just hang. I can run basic commands like ls no problem. But there does seem to be a problem as I cannot run either in my normal terminal. I didn't notice this before. 

edit: Update to most recent unstable, glxinfo works in native shell, but I can't enter the `nix-shell` inside the muvm repo it fails with:

```
<jemalloc>: Unsupported system page size
<jemalloc>: Unsupported system page size
<jemalloc>: Unsupported system page size
```

**Comment by waltmck** [comment]
A rootfs makes sense for FHS distros where programs look for specific locations in `/lib` to link against, but why is it necessary on NixOS? Shouldn't it be possible to pass the required `x86` libraries as `buildInputs` (using `patchelf` where necessary) like we do normally, and have the two architectures' libraries coexist in the Nix store? It should even be possible to make a `nix-ld` environment to run unpatched `x86` executables.

What is this approach missing?

**Comment by Sheepheerd** [comment]
@rowanG077 I believe your problem is coming from the unstable channel. Try building with stable and it should fix the problem. This is a problem with comping FEX on asahi. As for steam, I am able to launch it and start a game, but the game never opens and it just hangs. So I don't know where to go from there.

**Comment by stevesoltys** [comment]
I was able to get Steam running without compiling `muvm` from source, and it was actually pretty simple.

Instead of using `podman`, create a Fedora distrobox container and install all of the Asahi dependencies. Running the [Asahi steam wrapper](https://pagure.io/fedora-asahi/steam) directly from the distrobox container worked out of the box and allowed me to run games.

Here's a quick guide: https://gist.github.com/stevesoltys/af4c5c45317c543fa3b88da7d08a7218

Tested on:
1. MacBook Pro M2 Max
2. NixOS channel 25.11 (currently unstable)
3. OpenGL core profile version string: 4.6 (Core Profile) Mesa 25.2.4
4. Wayland / Sway / SDDM

**Comment by rowanG077** [comment]
@stevesoltys Thank  you for the guide. It doesn't work for me. I followed the installation guide and when I want to launch steam after a short time with all kinds of logs I get a new terminal popped up with:

```
You are missing the following 32-bit libraries, and Steam may not run:
libc.so.6
Press enter to continue:  
```

Pressing enter it continues logging and then gets stuck, the main window keeps displaying "Launching Steam."

<details>

<summary>Below is the complete log</summary>

```
📦[rowan.goemans@steam ~]$ steam
QApplication: invalid style override 'adwaita-dark' passed, ignoring it.
	Available styles: Windows, Fusion
Using default interface naming scheme 'v257'.
steam.sh[260]: Running Steam on fedora 42 64-bit
steam.sh[260]: STEAM_RUNTIME is enabled automatically
setup.sh[310]: Steam runtime environment up-to-date!
/home/rowan.goemans/Distrobox/steam/.local/share/Steam/ubuntu12_32/steam-runtime/usr/libexec/steam-runtime-tools-0/logger-0.bash: line 7:   333 Bus error                  rm -f "$fifo"
steam.sh[260]: Couldn't set up srt-logger, not logging to console-linux.txt
/home/rowan.goemans/Distrobox/steam/.local/share/Steam/steam.sh: line 821:   336 Bus error                  rm -fr "${STEAMROOT}/ubuntu12_64/steam-runtime-sniper/"
/home/rowan.goemans/Distrobox/steam/.local/share/Steam/steam.sh: line 822:   337 Bus error                  rm -f "${STEAMROOT}/ubuntu12_64/steam-runtime-sniper.lock"
/home/rowan.goemans/Distrobox/steam/.local/share/Steam/steam.sh: line 823:   338 Bus error                  rm -f "${STEAMROOT}/ubuntu12_64/steam-runtime-steamrt.lock"
steam.sh[260]: Error: You are missing the following 32-bit libraries, and Steam may not run:
libc.so.6
srt-logger[332]: E: Unable to report ready: Broken pipe
/home/rowan.goemans/Distrobox/steam/.local/share/Steam/steam.sh: line 62:   342 Aborted                    "${STEAM_ZENITY}" "$style" --text="$*" --width=450 2> /dev/null
steam.sh[260]: Steam client's requirements are satisfied
CProcessEnvironmentManager is ready, 6 preallocated environment variables.
[2025-10-19 16:02:08] Startup - updater built Oct  3 2025 02:26:15
[2025-10-19 16:02:08] Startup - Steam Client launched with: '/home/rowan.goemans/Distrobox/steam/.local/share/Steam/ubuntu12_32/steam' '-cef-force-occlusion'
10/19 16:02:09 minidumps folder is set to /tmp/dumps
10/19 16:02:09 Init: Installing breakpad exception handler for appid(steam)/version(1759461205)/tid(414)
Looks like steam didn't shutdown cleanly, scheduling immediate update check
[2025-10-19 16:02:09] Loading cached metrics from disk (/home/rowan.goemans/Distrobox/steam/.local/share/Steam/package/steam_client_metrics.bin)
[2025-10-19 16:02:09] Failed to load cached hosts file (File 'update_hosts_cached.vdf' not found), using defaults
[2025-10-19 16:02:09] Using the following download hosts for Public, Realm steamglobal
[2025-10-19 16:02:09] 1. https://client-update.steamstatic.com, /, Realm 'steamglobal', weight was 1, source = 'baked in'
[2025-10-19 16:02:09] Checking for update on startup
[2025-10-19 16:02:09] Checking for available updates...
CProcessEnvironmentManager is ready, 6 preallocated environment variables.
[2025-10-19 16:02:09] Downloading manifest: https://client-update.steamstatic.com/steam_client_ubuntu12
[2025-10-19 16:02:09] Manifest download: send request
[2025-10-19 16:02:09] Process started with command-line: '/home/rowan.goemans/Distrobox/steam/.local/share/Steam/ubuntu12_32/steam' '-child-update-ui' '-child-update-ui-socket' '16' '-cef-force-occlusion'
10/19 16:02:09 minidumps folder is set to /tmp/dumps
[2025-10-19 16:02:09] Using update UI: console
10/19 16:02:09 Init: Installing breakpad exception handler for appid(steam)/version(0)/tid(417)
[2025-10-19 16:02:09] Create window
[2025-10-19 16:02:09] Set percent complete: 0
[2025-10-19 16:02:09] Set status message: Checking for available updates...
[  0%] Checking for available updates...
[2025-10-19 16:02:09] Set percent complete: -1
[2025-10-19 16:02:09] Manifest download: waiting for download to finish
[2025-10-19 16:02:09] Manifest download: finished
[2025-10-19 16:02:09] Download skipped: /steam_client_ubuntu12 version 1759461205, installed version 1759461205, existing pending version 0
[2025-10-19 16:02:09] Nothing to do
[2025-10-19 16:02:09] Verifying installation...
[2025-10-19 16:02:09] Verifying all executable checksums
[2025-10-19 16:02:09] Set percent complete: -1
[2025-10-19 16:02:09] Set status message: Verifying installation...
[----] Verifying installation...
[2025-10-19 16:02:10] Verification complete
UpdateUI: skip show logo
[2025-10-19 16:02:10] Destroy window

Steam logging initialized: directory: /home/rowan.goemans/Distrobox/steam/.local/share/Steam/logs

[2025-10-19 16:02:10] ProcessNextMessage: socket disconnected
[2025-10-19 16:02:10] No more messages are expected - exiting
pcilib: Cannot open /proc/bus/pci
lspci: Cannot find any working access method.
XRRGetOutputInfo Workaround: initialized with override: 0 real: 0xf4b47f90
XRRGetCrtcInfo Workaround: initialized with override: 0 real: 0xf4b46670
10/19 16:02:12 minidumps folder is set to /tmp/dumps
10/19 16:02:12 Init: Installing breakpad exception handler for appid(steamsysinfo)/version(1759461205)/tid(428)
Running query: 1 - GpuTopology
Response: gpu_topology {
  gpus {
    id: 1
    name: "Apple M2 Max (G14C B1)"
    vram_size_bytes: 33655095296
    driver_id: k_EGpuDriverId_Unknown
    driver_version_major: 25
    driver_version_minor: 2
    driver_version_patch: 4
    luid: 0
  }
  default_gpu_id: 1
}

Exit code: 0
Saving response to: /tmp/steamalJJea - 48 bytes
pcilib: Cannot open /proc/bus/pci
lspci: Cannot find any working access method.
Steam Runtime Launch Service: starting steam-runtime-launcher-service
Steam Runtime Launch Service: steam-runtime-launcher-service is running pid 460
bus_name=com.steampowered.PressureVessel.LaunchAlongsideSteam
[2025-10-19 16:04:13] Background update loop checking for update. . .
[2025-10-19 16:04:13] Checking for available updates...
[2025-10-19 16:04:13] Downloading manifest: https://client-update.steamstatic.com/steam_client_ubuntu12

thread '<unnamed>' panicked at src/guest/bridge/x11.rs:249:52:
range end index 14 out of range for slice of length 8
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
x11bridge thread crashed, x11 passthrough will no longer function
```

</details>

I'm on M2 Max and hyprland.

**Comment by stevesoltys** [comment]
@rowanG077 I would recommend trying using the exact configuration I provided in the guide. I've updated it to include things like kernel version and `nixos-apple-silicon` version.

I moved away from Hyprland before trying any of this since I was having tons of issues with it on Asahi.

**Comment by rbraley** [comment]
I had to make a few changes to start it from within distrobox but it works for me in Plasma 6:
unset QT_PLUGIN_PATH
unset QML2_IMPORT_PATH
PATH="/bin:$PATH" LD_LIBRARY_PATH=$NIX_LD_LIBRARY_PATH steam

**Comment by themg95** [comment]
are there any updates on if we can use muvm without a fedora container now?

**Comment by slendidev** [comment]
I got muvm+fex to kinda work, however I cannot for the life of me get graphics acceleration. Basic X11 applications like xeyes work though. This is what I did: https://paste.slendi.dev/hfkG8S.txt. There are multiple issues:

1. You won't get proper support when it comes to things like CLIs like bash, leading to things like the prompt missing, or neovim not doing any of the TUI related stuff. I found this PR while looking for a solution: https://github.com/AsahiLinux/muvm/pull/157
2. No graphics acceleration. EGL fails to initialize and get stuff like this:
```
glxinfo -B
Error: couldn't find RGB GLX visual or fbconfig
name of display: :1
```
3. Relative paths can't be executed.

Number 1 and 3 may be addressed possibly through a custom daemon in the guest, but for number 2 maybe someone who actually knows what they are doing with graphics could help.

**Comment by slendidev** [comment]
@waltmck I have tried that and could only get it to work properly on 4k page size using this: https://paste.slendi.dev/CTm8rg.txt, however the M chips only having 16k page size support in their IOMMU considerably complicates this, since you have to run a microVM. That's where a lot of the problems arise from.

**Comment by waltmck** [comment]
> [@waltmck](https://github.com/waltmck) I have tried that and could only get it to work properly on 4k page size using this: https://paste.slendi.dev/CTm8rg.txt, however the M chips only having 16k page size support in their IOMMU considerably complicates this, since you have to run a microVM. That's where a lot of the problems arise from.

Unfortunately FEX [decided not to support](https://github.com/FEX-Emu/FEX/issues/1921) 16K pages. There used to be many programs that were working (including AAA games like Red Dead Redemption) on 16K pages, but these are broken since FEX 2508. It's especially tragic because NT memory allocations are 64K-aligned, so in almost all cases arm64ec wine should just work with minimal effort (the only exception are apps that allocate a page and later change permission of a 4K chunk of that page; FEX would need to implement actual page size emulation in order to support that behavior).

I also don't think `muvm` is as promising as I used to. I guess it's okay for gaming, but (aside from the memory usage) we would need to constantly fight the kernel to support things like forwarding domain sockets across VMs or dbus/polkit. KVM is just not designed for integrating a guest graphical application into a modern Linux desktop, and the workarounds (`passt`) seem fragile and hard to maintain (as an example, `atuin` breaks inside `muvm` since the socket isn't forwarded to the guest). All of this isn't insurmountable, but development on `muvm` seems to have stalled with the last nontrivial commit being more than 4 months ago.

In the long-term, I think that the only elegant way to solve this is to add heterogeneous page size support to the kernel (as evidence, this is what Apple did with XNU at significant cost). In the medium term, I hope that FEX will reconsider their position on large page sizes and integrate with WINE to support emulating x86 Windows binaries. For the short term, I've just switched to using `box64`, which works fine for a few games I've tried with better performance than I've been able to get from FEX (due to being able to get thunking to work) and without `muvm`.

**Comment by neobrain** [comment]
See https://github.com/sm-idk/asahi-steam for a FEX+muvm+Asahi+NixOS setup that's at least been working for the author.

**Comment by flokli** [comment]
Emulating x86 on an aarch64 system sounds like a generic problem, not necessarily limited to asahi. There's other arm64 hardware out there that can run in 16K pages mode, and until the kernel gets heterogenous page size support we might not only need this to emulate x86 on Asahi machines.

So I'd strongly encourage to move this discussion upstream. Maybe some, if not all code in that flake can be absorbed into nixpkgs.

**Comment by ImUrX** [comment]
It would have been nice to make a new issue upstream when this one got closed. I don't know if I chose the correct one to make the discussion happen but I passed the info there.

**Comment by rowanG077** [comment]
I’ve been working on [nix-x86-on-aarch64](https://github.com/rowanG077/nix-x86-on-aarch64), which I think is a significant upgrade over [steam-asahi](https://github.com/sm-idk/steam-asahi):

- The NixOS module registers both x86_64 and i686 binfmt handlers, so ordinary x86 Linux programs can be launched directly.
- The guest userspace is assembled entirely from nixpkgs. There is no Fedora rootfs download or runtime rootfs fetcher.
- NixOS provides an x86pkgs / x86packages package set, so commands like nix-shell -p x86pkgs.hello and x86 Wine work through the normal Nix workflow.
- It supports generic x86 commands, Wine, and Steam, with direct FEX on 4 KiB hosts and muvm on larger-page hosts.

**Comment by waltmck** [comment]
I've been working on [propnix](https://github.com/waltmck/propnix), which aims to package specific proprietary packages along with specific WINE/FEX/box64 tweaks to get them to work on their target platform. So far I have the following games confirmed running on Asahi:

- Baby Steps
- Baldurs Gate 3
- Don't Starve
- Factorio (including Space Age)
- Fallout New Vegas
- Hollow Knight
- Hollow Knight: Silksong
- Iron Lung
- Iron Nest
- Kerbal Space Program
- No Man's Sky
- Outlast
- Outlast 2
- Papers Please
- Prison Architect
- Skyrim Special Edition
- Stellaris

These use a variety of backends and tricks in order to get them to work, and propnix has a full module system to support declarative configuration. I also packaged fetchers from GoG and Steam, so running these is as simple as `nix run .#factorio` after credentials are set up. DLC is also fully supported (e.g. `nix run .#factorio.withAllDlc`).

Adding a game is generally quite easy, I am open to PRs to try to get more stuff working. See the examples in `pkgs/games`.
